File: PythonSolution/SubstringwithConcatenationofAllWords.py
class Solution:
    def findSubstring(self, s: str, words):
        if not words:
            return []
        count = {}
        for i in range(len(words)):
            count[words[i]] = count.get(words[i], 0) + 1
        
        wordsLen = len(words)
        wordLen = len(words[0])
        res = []
        for i in range(len(s) - wordsLen * wordLen + 1):
            dic = {}
            j = 0
            while(j < wordsLen):
                subS = s[i + j * wordLen : i + (j + 1) * wordLen]
                dic[subS] = dic.get(subS, 0) + 1
                if(dic[subS] > count.get(subS, 0)):
                    break
                j += 1
            if j == wordsLen:
                res.append(i)
        return res

File: Assets/Second/SubstringwithConcatenationofAllWords.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class SubstringwithConcatenationofAllWords : MonoBehaviour
{
    // Start is called before the first frame update
    void Start()
    {
        
    }

    // Update is called once per frame
    void Update()
    {
        
    }

    /**
    https://leetcode.com/problems/substring-with-concatenation-of-all-words/

    使用两个map来进行判断，第一个map用于记录words数组中所有单词的出现次数
    第二个map用于每次迭代当前迭代的字母开始单词出现的次数
     */
    public IList<int> FindSubstring(string s, string[] words) {
        if(words == null || words.Length == 0) return new List<int>();
        Dictionary<string, int> count = new Dictionary<string, int>();
        
        foreach(string str in words){
            if(count.ContainsKey(str)){
                count[str]++;
            }else{
                count.Add(str, 1);
            }
        }
        
        int n = s.Length;
        int sum = words.Length;
        int len = words[0].Length;
        IList<int> res = new List<int>();
        for(int i = 0; i <= n - sum * len; i++){
            Dictionary<string, int> dic = new Dictionary<string, int>();
            int j = 0;
            while(j < sum){
                string str = s.Substring(i + j * len, len);
                if(dic.ContainsKey(str)){
                    dic[str]++;
                }else{
                    dic.Add(str, 1);
                }
                
                if(!count.ContainsKey(str)){
                    break;
                }
                
                if(dic[str] > count[str]){
                    break;
                }
                
                j++;
            }
            
            if(j == sum){
                res.Add(i);
            }
        }
        
        return res;
    }
}
